node/ChainwebNode.hs

```diff
@@ -299,20 +305,32 @@ node conf logger = do
     dbBaseDir <- getDbBaseDir conf
     when (_nodeConfigResetChainDbs conf) $ removeDirectoryRecursive dbBaseDir
     rocksDbDir <- getRocksDbDir conf
+    rocksDbCheckpointDir <- getRocksDbCheckpointDir conf
     pactDbDir <- getPactDbDir conf
     withRocksDb rocksDbDir $ \rocksDb -> do
         logFunctionText logger Info $ "opened rocksdb in directory " <> sshow rocksDbDir
-        withChainweb cwConf logger rocksDb pactDbDir (_nodeConfigResetChainDbs conf) $ \cw -> mapConcurrently_ id
-            [ runChainweb cw
-              -- we should probably push 'onReady' deeper here but this should be ok
-            , runCutMonitor (_chainwebLogger cw) (_cutResCutDb $ _chainwebCutResources cw)
-            , runQueueMonitor (_chainwebLogger cw) (_cutResCutDb $ _chainwebCutResources cw)
-            , runRtsMonitor (_chainwebLogger cw)
-            , runBlockUpdateMonitor (_chainwebLogger cw) (_cutResCutDb $ _chainwebCutResources cw)
-            ]
+        bracket (newIORef (Just rocksDb)) (flip writeIORef Nothing) $ \ref -> do
+            installHandlerCross sigUSR1 (const $ makeCheckpoint rocksDbCheckpointDir ref)
+            withChainweb cwConf logger rocksDb pactDbDir (_nodeConfigResetChainDbs conf) $ \cw -> mapConcurrently_ id
+                [ runChainweb cw
+                  -- we should probably push 'onReady' deeper here but this should be ok
+                , runCutMonitor (_chainwebLogger cw) (_cutResCutDb $ _chainwebCutResources cw)
+                , runQueueMonitor (_chainwebLogger cw) (_cutResCutDb $ _chainwebCutResources cw)
+                , runRtsMonitor (_chainwebLogger cw)
+                , runBlockUpdateMonitor (_chainwebLogger cw) (_cutResCutDb $ _chainwebCutResources cw)
+                ]
   where
     cwConf = _nodeConfigChainweb conf
 
+makeCheckpoint :: FilePath -> IORef (Maybe RocksDb) -> IO ()
+makeCheckpoint checkpointDir rdbRef = 
+    readIORef rdbRef >>= \case
+        Nothing -> pure ()
+        -- 0 ~ never flush WAL log before checkpoint, to avoid making extra work
```

Comment:
Do you know how often the WAL log is flushed normally? Assuming that it's done often, this should be fine. Otherwise, we may consider flushing.

Writing the snapshot can only be triggered by a node admin and probably won't happen too often, so some performance overhead is probably fine.

Reply:
It's flushed when it gets to 4MB IIRC. This makes sense to me, avoiding extra catchup is probably more important than a momentary hiccup.

Reply:
4MB is probably around a few thousand blocks. I think, it's fine for now to not flush. Maybe we could do a follow up PR that addresses it. We could use another signal for flushing or add an (optional) endpoint on the service API that would take parameters (e.g. snapshot name and flushing)

Reply:
Oh, just noticed that you changed it. It's fine either way.